programs/lib/flexpret_exceptions.c

```diff
@@ -41,66 +52,52 @@ static const char *exception_to_str(const uint32_t cause) {
     case EXE_CAUSE_EXT_INTERRUPT6: return "External interrupt 6";
     case EXE_CAUSE_TIMER_INTERRUPT: return "Timer interrupt";
     case EXC_CAUSE_INTERRUPT_EXPIRE: return "Interrupt expire";
-    case EXC_CAUSE_EXTERNAL_INT: return "External int"; // FIXME: int = interrupt? 
+    case EXC_CAUSE_EXTERNAL_INT: return "External interrupt";
     default: return "Unknown exception code";
     }
 }
 
-static void fp_exception_handler(void) {
+void fp_exception_handler(void) {
     int cause = read_csr(CSR_CAUSE);
-    
+    uint32_t hartid = read_hartid();
+
     if (cause == EXC_CAUSE_EXTERNAL_INT) {  
-        if(ext_int_handler) ext_int_handler();
+        if(ext_int_handler[hartid]) ext_int_handler[hartid]();
     } else if (cause == EXC_CAUSE_INTERRUPT_EXPIRE) {
-        if(ie_int_handler) ie_int_handler();
+        if(ie_int_handler[hartid]) ie_int_handler[hartid]();
     } else if (cause == EXC_CAUSE_EXCEPTION_EXPIRE) {
-        if(ee_int_handler) ee_int_handler();
+        if(ee_int_handler[hartid]) ee_int_handler[hartid]();
     } else {
-        printf("Exception occured: %i, %s\n", cause, exception_to_str(cause));
-        assert(false, "Exception not handled");
+        fp_assert(false, "Exception not handled: %i, %s\n", cause, exception_to_str(cause));
     }
+
+    // Call the function to load the thread's context
+    // We mark it with attribute noretun because the function will not return
+    // to fp_exception_handler, but instead to where the exception occurred.
+    void thread_ctx_switch_load(void) __attribute__((noreturn));
```

Comment:
I havent seen this before. Is it a forward declaration of a function, within another function? Couldnt `thread_ctx_switch_load` be defined with that attribute in the original declaration?

Reply:
There is no original declaration of the function in C because it is implemented in assembly (see ctx_switch.S)

It could of course be declared at file scope or in a header file, but since `fp_exception_handler` is the only function that uses it, I added it here. I think declaring it inside a function is just a matter of scoping - it does not make a difference to declare it at file level scope. But I guess it is a bit confusing to declare it inside the function, so I can move it outside.

The advantage of having it declared inside the function is that it becomes clear that it is not used anywhere else. A disadvantage is that it might cause some confusion. 

For the record: The `__attribute_((noreturn))` is not at all necessary; it just slightly reduces code size because it removes the "return to caller" part of the function.

Reply:
OK, I would prefer putting this at the top of the file, to avoid any confusion :)